src/project/internal/notationproject.cpp

```diff
@@ -830,33 +873,39 @@ void NotationProject::markAsSaved(const io::path_t& path)
     //! NOTE: order is important
     m_isNewlyCreated = false;
 
-    m_needSaveNotificationBlocked = true;
+    setNeedSave(false);
+
+    setPath(path);
 
-    // Mark engraving project as saved
-    m_masterNotation->setSaved(true);
+    m_hasNonUndoStackChanges = false;
 
-    // Mark audio settings as saved
-    m_projectAudioSettings->markAsSaved();
+    m_masterNotation->notation()->undoStack()->stackChanged().notify();
+}
 
-    // Mark view settings as saved
-    m_masterNotation->notation()->viewState()->markAsSaved();
-    for (IExcerptNotationPtr excerpt : m_masterNotation->excerpts().val) {
-        excerpt->notation()->viewState()->markAsSaved();
+void NotationProject::setNeedSave(bool needSave)
+{
+    mu::engraving::MasterScore* score = m_masterNotation->masterScore();
+    if (!score) {
+        return;
     }
 
-    m_needSaveNotificationBlocked = false;
-    m_needSaveNotification.notify();
+    setNeedAutoSave(needSave);
```

Comment:
Maybe this logic is still not perfect; what about the following situation?
1. User saves file
2. User performs action X -> marked as "need save" and "need autosave"
3. Wait until autosave does its work -> marked as "not need autosave"
4. User performs action Y -> marked as "need autosave"
5. User undoes action Y -> expected: marked as "not need autosave"; actual: still marked as "need autosave"

It's quite an edge case, so I'm not sure if we really need to worry about it.

Reply:
This looks like a very edge case that would probably require a significant over-complication of the implementation (unless I'm missing something)

Reply:
There might be a semi-easy fix: inside Libmscore, each state of the UndoStack gets an ID. If you expose this ID via INotationUndoStack, you can use it in NotationProject: you just store the state ID of the last save and of the last autosave. Once you have that info, things are easy:
```c++
needSave == (m_lastSaveUndoStackStateId != undoStack()->currentStateId()) || m_hasNonUndoStackChangesSinceLastSave;
needAutosave == (m_lastAutosaveUndoStackStateId != undoStack()->currentStateId()) || m_hasNonUndoStackChangesSinceLastAutosave;
```
You only need to be careful with the value of `m_lastSaveUndoStackStateId` and `m_lastAutosaveUndoStackStateId` when the score has never been saved yet or when the undo stack is empty. 

Anyway, perhaps that's too much of a complication, given how little value it brings.